server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexAliasesServiceTests.java

```diff
@@ -675,6 +679,95 @@ public void testDataStreamAliasesWithWriteFlag() {
+        );
+        IndicesAliasesClusterStateUpdateRequest removeAliasRequest = new IndicesAliasesClusterStateUpdateRequest(
+            List.of(new AliasAction.Remove(index, "test", true))
+        );
+
+        AtomicInteger counter = new AtomicInteger();
+        ActionListener<AcknowledgedResponse> listener1 = new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                assertThat(acknowledgedResponse.isAcknowledged(), equalTo(true));
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                fail("Unexpected failure " + e.getMessage());
+            }
+        };
+        ActionListener<AcknowledgedResponse> listener2 = new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                assertThat(acknowledgedResponse.isAcknowledged(), equalTo(true));
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                fail("Unexpected failure " + e.getMessage());
+            }
+        };
+        ActionListener<AcknowledgedResponse> listener3 = new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                assertThat(acknowledgedResponse.isAcknowledged(), equalTo(true));
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                fail("Unexpected failure " + e.getMessage());
+            }
+        };
+        ActionListener<AcknowledgedResponse> listener4 = new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                assertThat(acknowledgedResponse.isAcknowledged(), equalTo(true));
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                fail("Unexpected failure " + e.getMessage());
+            }
+        };
+        ClusterState after = ClusterStateTaskExecutorUtils.executeAndAssertSuccessful(
+            before,
+            executor,
+            List.of(
+                new MetadataIndexAliasesService.ApplyAliasActions(service, addAliasRequest, listener1),
+                // Repeat the same change to ensure that the cluster state version won't increase
+                new MetadataIndexAliasesService.ApplyAliasActions(service, addAliasRequest, listener2),
+                new MetadataIndexAliasesService.ApplyAliasActions(service, removeAliasRequest, listener3),
+                new MetadataIndexAliasesService.ApplyAliasActions(service, addAliasRequest, listener4)
+            )
+        );
+
+        waitUntil(() -> counter.get() == 4);
```

Comment:
Nit: no real need for 4 different listeners here, nor a busy-wait for completion:

```suggestion
        final var latch = new CountDownLatch(4);
        final var listener = new ActionListener<AcknowledgedResponse>() {
            @Override
            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
                assertTrue(acknowledgedResponse.isAcknowledged());
                latch.countDown();
            }

            @Override
            public void onFailure(Exception e) {
                fail("Unexpected failure " + e.getMessage());
            }
        };
        ClusterState after = ClusterStateTaskExecutorUtils.executeAndAssertSuccessful(
            before,
            executor,
            List.of(
                new MetadataIndexAliasesService.ApplyAliasActions(service, addAliasRequest, listener),
                // Repeat the same change to ensure that the cluster state version won't increase
                new MetadataIndexAliasesService.ApplyAliasActions(service, addAliasRequest, listener),
                new MetadataIndexAliasesService.ApplyAliasActions(service, removeAliasRequest, listener),
                new MetadataIndexAliasesService.ApplyAliasActions(service, addAliasRequest, listener)
            )
        );
        assertTrue(latch.await(10, TimeUnit.SECONDS));
```